crates/circuit/src/dag_circuit.rs

```diff
@@ -239,11 +199,14 @@ pub struct DAGCircuit {
 
     // Python modules we need to frequently access (for now).
     control_flow_module: PyControlFlowModule,
-    vars_info: HashMap<String, DAGVarInfo>,
+
+    identifier_info: HashMap<String, DAGIdentifierInfo>,
+
+    // Indexed by DAGVarType as usize
     vars_by_type: [Py<PySet>; 3],
 
-    captured_stretches: IndexMap<String, Py<PyAny>, RandomState>,
-    declared_stretches: IndexMap<String, Py<PyAny>, RandomState>,
+    captured_stretches: Py<PySet>,
+    declared_stretches: Py<PyList>,
```

Comment:
After #14069 we're not storing bare python objects in the dag they're all wrapped in either an `Option<>` or some other struct so we can defer the creation to a context where we have the py token to ensure we have a path to create a dag without the pytoken. So I don't think this will work when you rebase to main. We can wrap it in an `Option`, `OnceLock` or something else and maybe just make a `ContainedStretches` struct, or something similar to manage the lazy creation of them.